### PR TITLE
Fixes of Hierarchical ORTModule and ORTModule PythonOp

### PR DESCRIPTION
**Pull request by iK1D**
The PR applies some fixes to Hierarchical ORTModule and ORTModule PythonOp.

For Hierarchical ORTModule:
- Don't wrap module if the caller is to call other function instead of forward() function
- Support single module instance is call multiple times with different types of inputs
- Check if module can be warped from top to bottom instead of from bottom to top

For ORTModule PythonOp:
- Add env variable control to allow using torch.utils.checkpoint.CheckpointFunction
- Add env variable control to skip register some autograd functions so that there is no conflict for some models.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py [hunk 1]
```diff
@@ -15,6 +18,40 @@
 )
 
 
+class IteratedORTModule(torch.nn.Module):
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py:21]: Do you think we could instead provide a caching mechanism inside ortmodule that caches the exported graphs/inference session (and other dependencies) on the input types hash, and then based on that executes the graph.
I think users might also benefit if ortmodule is able to support multiple types without needing to re-export the model.

We can do this later considering this might take longer, but just wanted to hear your thoughts on this.

Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py:21]: If this is internal only, can we add `_` prefix to its name?

File: orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py [hunk 2]
```diff
@@ -69,38 +108,30 @@ def record_args(module, args):
                 module_arg_pool[module] = [args]
 
         # Recursively hook "record_args" to module and all its sub-modules.
-        # The function "record_args" records the inputs for each nn.Module,
-        # and later we will try exporting those nn.Module's with their recorded
-        # inputs.
+        # The function "record_args" records the inputs for each nn.Module and later we will try exporting
+        # those nn.Module's with their recorded inputs.
+        # NOTE that if a module is not called from forward(), it will fail to be captured by this hook.
```
[the remark this answers is not in the document]
Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py:113]: The MoE model has such usage, and there is sample usage in the new added UT.

File: orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py (hunk 1 above)
Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py:21]: We can think about it. Hash to graph is similar to hash to ORTModule, because each time we still need to export new graph and build gradients. We can think about the easiest way to implement.

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py [hunk 3]
```diff
@@ -238,7 +249,10 @@ def _post_process_after_export(exported_model, enable_custom_autograd_function,
 
 def _post_process_enabling_autograd_fallback(exported_model):
     registered_name_mappings = {}
+    skipped_autograd_function_list = ortmodule._defined_from_envvar("SKIPPED_AUTOGRAD_FUNCTIONS", "").split(",")
     for kclass in torch.autograd.Function.__subclasses__():
+        if full_name(kclass) in skipped_autograd_function_list:
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:254]: full_name() is did similar thing with line 259 (considering autograd.Function will not be python built-in module). You can keep either of them.

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py [hunk 4]
```diff
@@ -66,7 +74,10 @@ def _export_pt_1_10(g, n, *args, **kwargs):
     """
     try:
         name = kwargs["name"]
-        if name in BANNED_AUTOGRAD_FUNCTION_NAMES:
+        if name in BANNED_AUTOGRAD_FUNCTION_NAMES and (
+            not ortmodule._defined_from_envvar("ALLOW_AUTOGRAD_CHECKPOINT", 0)
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:78]: I believe our env variable name usually starts with ORTModule_?

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py [hunk 5]
```diff
@@ -42,6 +43,13 @@
 }
 
 
+def full_name(klass):
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:46]: >full_name [](http://example.com/codeflow?start=4&length=9)

_full_name()

**Review by baijumeswani — APPROVED** [review]
